### PR TITLE
Move back::link and debuginfo::type_names to cg ssa

### PR DESCRIPTION
**Pull request by bjorn3**
r? @eddyb

### CODE REVIEW

**Comment by bjorn3** [comment]
Rebased

**Comment by bjorn3** [comment]
Rebased

**Comment by bjorn3** [comment]
Rebased (again)

File: src/librustc_codegen_ssa/lib.rs [hunk 1]
```diff
@@ -66,6 +66,7 @@ pub struct ModuleCodegen<M> {
     pub kind: ModuleKind,
 }
 
+pub const METADATA_FILENAME: &str = "rust.metadata.bin";
 pub const RLIB_BYTECODE_EXTENSION: &str = "bc.z";
```
Comment (eddyb) [inline src/librustc_codegen_ssa/lib.rs]: This is a problem, because `bc` is LLVM-specific.

File: src/librustc_codegen_ssa/back/link.rs [hunk 2]
```diff
@@ -1,21 +1,40 @@
 /// For all the linkers we support, and information they might
 /// need out of the shared crate context before we get rid of it.
```
Comment (eddyb) [inline src/librustc_codegen_ssa/back/link.rs]: cc @alexcrichton @michaelwoerister I feel like this module would be better suited for another crate, say `rustc_codegen_link` or `rustc_link`(`er`), what do you think?

File: src/librustc_codegen_ssa/lib.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc_codegen_ssa/lib.rs]: But feel free to file an issue and leave it for later.

**Comment by eddyb** [comment]
r=me with or without comments addressed

File: src/librustc_codegen_ssa/lib.rs (hunk 1 above)
Comment (bjorn3) [reply src/librustc_codegen_ssa/lib.rs]: What is the convention on those kind of issues? File it right now and add a comment here referencing it?

File: src/librustc_codegen_ssa/back/link.rs (hunk 2 above)
Comment (bjorn3) [reply src/librustc_codegen_ssa/back/link.rs]: Who should depend on who? `CrateInfo` and `CodegenResults` are both needed by `rustc_codegen_ssa` and the linker code.

**Comment by bjorn3** [comment]
Rebased

**Comment by bjorn3** [comment]
Rebased

**Comment by bjorn3** [comment]
Rebased

**Comment by Xanewok** [comment]
@bjorn3 maybe this needs another rebase?
```
...
Auto-merging Cargo.lock
CONFLICT (content): Merge conflict in Cargo.lock
```
@bors delegate+


I experienced something similar in https://github.com/rust-lang/rust/pull/60053#issuecomment-484936038.
@pietroalbini is it because how homu manages branches specifically or it uses some non-default merge strategy? (and why is Cargo.lock treated as a binary file?)

**Comment by philipc** [comment]
@Xanewok See the Cargo.lock entry in  .gitattributes, and https://github.com/rust-lang/highfive/issues/153

**Comment by Xanewok** [comment]
Ah, I didn't know this was enforced on the merge level; thanks @philipc!

**Comment by bjorn3** [comment]
Rebased

@bors r=eddyb

**Pull request merged**